File: golang/yaml/decoder.go
package yaml

import (
	"encoding/json"
	"fmt"
	"github.com/goccy/go-yaml"
)

type Decoder interface {
	Decode(*DecodeInput) *DecodeOutput
}
type decoder struct{}

func NewDecoder() Decoder {
	return &decoder{}
}
func (u *decoder) Decode(input *DecodeInput) *DecodeOutput {
	b, err := yaml.YAMLToJSON([]byte(input.Yaml))
	if err != nil {
		return &DecodeOutput{
			IsError:      true,
			ErrorMessage: fmt.Sprintf("fail to convertFromGo yaml to json: %+v", err),
		}
	}
	var v any
	if err := json.Unmarshal(b, &v); err != nil {
		return &DecodeOutput{
			IsError:      true,
			ErrorMessage: fmt.Sprintf("fail to unmarshal json: %+v", err),
		}
	}
	return &DecodeOutput{Value: convertFromGo(v)}
}

func convertFromGo(v any) *Value {
	switch v := v.(type) {
	default:
		panic(fmt.Sprintf("unexpected type %T", v))
	case nil:
		return &Value{Type: Type_TYPE_NULL}
	case bool:
		return &Value{Type: Type_TYPE_BOOL, Bool: v}
	case float64:
		return &Value{Type: Type_TYPE_NUM, Num: v}
	case string:
		return &Value{Type: Type_TYPE_STR, Str: v}
	case []interface{}:
		arr := []*Value{}
		for _, elem := range v {
			arr = append(arr, convertFromGo(elem))
		}
		return &Value{Type: Type_TYPE_ARR, Arr: arr}
	case map[string]interface{}:
		obj := map[string]*Value{}
		for key, value := range v {
			obj[key] = convertFromGo(value)
		}
		return &Value{Type: Type_TYPE_OBJ, Obj: obj}
	}
}

File: typescript/src/yaml/decoder.ts
import {create} from "@bufbuild/protobuf";
import {DecodeInput, DecodeOutput, DecodeOutputSchema} from "./decoder_pb.js";
import YAML from 'yaml'
import {Type, Value, ValueSchema} from "./value_pb.js";

export class Decoder {
    decode(input: DecodeInput): DecodeOutput {
        try {
            const o = YAML.parse(input.yaml);
            const v = this.convertFromJS(o);
            return create(DecodeOutputSchema, {value: v});
        } catch (e) {
            const msg = e instanceof Error ? e.message : JSON.stringify(e);
            return create(DecodeOutputSchema, {
                isError: true,
                errorMessage: msg,
            });
        }
    }

    private convertFromJS(v: unknown): Value {
        if (v === null) {
            return create(ValueSchema, {type: Type.NULL});
        } else if (typeof v === "boolean") {
            return create(ValueSchema, {type: Type.BOOL, bool: v});
        } else if (typeof v === "number") {
            return create(ValueSchema, {type: Type.NUM, num: v});
        } else if (typeof v === "string") {
            return create(ValueSchema, {type: Type.STR, str: v});
        } else if (Array.isArray(v)) {
            return create(ValueSchema, {type: Type.ARR, arr: v.map(this.convertFromJS)});
        } else if (typeof v === "object") {
            return create(ValueSchema, {
                type: Type.OBJ,
                obj: Object.fromEntries(Object.entries(v).map(([k, v]) => [k, this.convertFromJS(v)]))
            });
        }
        throw new Error("unexpected value type");
    }
}
